### DISCUSSION TITLE
Load HTML File Programmatically

### DISCUSSION THREAD

**Discussion by amishtp1407** [tauri-apps/tauri#11731] [Q&A]
I am facing issues in loading different HTML files. Currently I am using the Vite Dev Server, to load different HTML files, but this will stop working once I bundle the project. How can I load different HTML files into my webview, without a use of localhost server.

```
async fn login(server: &str, username: &str, password: &str, window: tauri::Window) ->  Result<serde_json::Value, String> {
    println!("{} {} {}", server, username, password);
    match manager::authenticate(server, username, password).await {
        Ok(res) => {
            if res.valid {
                println!("{}", res.message);
                let webview = window.get_webview_window("main");
                let mut url = Url::parse("http://localhost:1420/tunnel");
                if res.code == 302 {
                    url = Url::parse("http://localhost:1420/mfa");
                }

                let _ = webview.unwrap().navigate(url.unwrap());
                Ok(json!({"message": "Authentication successful, loading login page."}))
            } else {
                Ok(json!(res))
            }
        }

        Err(_) => {
            let err = DataResponse {
                code: 400,
                valid: false,
                message: "Failed to connect with Console".to_string()
            };

            Ok(json!(err))
        }
    }
}
```

**Comment by FabianLars** [comment]
Maybe https://docs.rs/tauri/2.0.0/tauri/struct.Builder.html#method.register_uri_scheme_protocol or its async variant below that? This is what we use for the `tauri://` protocol which serves the bundled frontend files.

**Reply by amishtp1407** [reply]
Hi Fabian,

Thank you for the answer. 

Just have a doubt! std::fs::read(path) from which directory does it read the file? and can this method be used for html files that have typescript files, or will I have to use frontend navigation method.

**Reply by FabianLars** [reply]
> std::fs::read(path) from which directory does it read the file?

If it's a relative path then the path will be relative to the exe location in production (typically read only btw), and i think relative to the src-tauri dir in development.
If it's an absolute path then that of course doesn't matter.

> and can this method be used for html files that have typescript files, or will I have to use frontend navigation method.

Browsers don't understand typescript so they must be converted to javascript first. Similar story if those files require npm modules, those likely have to go through a bundler too to work.
If that's handled then you may also have to change to imports in the html file to load from the custom protocol specifically (probably not though), otherwise the browser may request them from the built-in tauri protocol. But i'm not 100% sure about this so maybe try it out first.

**Reply by amishtp1407** [reply]
understood Fabian, I am thinking of switching to a React based Frontend, so that I  will implement a state based component rendering, I think this will handle the issue of loading by programmatically, I just need to setup for event listening at front end. What do you think of the above method?

**Reply by FabianLars** [reply]
> What do you think of the above method?

The state based component rendering? I think that it may be more straight forward yes.

One thing about your old approach though. You said you're using vite for this, but this should also mean that it should work in prod as long as vite writes those html files into the frontendDist. Of course you'd have to adapt the urls to use `tauri://localhost/` and `http://tauri.localhost/` instead of localhost:1420 in prod.

**Reply by amishtp1407** [reply]
thanks Fabian, I will first try with the old approach first, since the code is ready, if it works, then amazing! otherwise React is there
